lib/elixir/lib/kernel.ex

```diff
@@ -1599,19 +1599,63 @@ defmodule Kernel do
   # Extracts concatenations in order to optimize many
   # concatenations into one single clause.
   defp extract_concatenations({:<>, _, [left, right]}) do
-    [wrap_concatenation(left) | extract_concatenations(right)]
+    [wrap_concatenation(left, :left) | extract_concatenations(right)]
   end
 
   defp extract_concatenations(other) do
-    [wrap_concatenation(other)]
+    [wrap_concatenation(other, :right)]
   end
 
-  defp wrap_concatenation(binary) when is_binary(binary) do
+  defp wrap_concatenation(binary, _side) when is_binary(binary) do
     binary
   end
 
-  defp wrap_concatenation(other) do
-    {:::, [], [other, {:binary, [], nil}]}
+  defp wrap_concatenation(other, side) do
+    {:::, [],
+     [
+       {:expand_concatenation_argument, [context: Elixir, import: Kernel], [other, side]},
+       {:binary, [], nil}
+     ]}
+  end
+
+  @doc false
+  # Used internally to check the arguments expansion of <> operator.
+  # This function is private and must be used only internally.
+  defmacro expand_concatenation_argument(other, side) do
+    expand =
+      case bootstrapped?(Macro) do
+        true -> &Macro.expand(&1, __CALLER__)
+        false -> & &1
+      end
+
+    case {expand.(other), __CALLER__.context, side} do
```

Comment:
I like this direction a lot!

However, Macro.expand can change the environment but because  we cannot tell Elixir compiler that the environment changed, information can get lost along the way.

This is, however, not a problem inside matches because they can't do anything to change the environment anyway (such as defining aliases, requires, etc).

So my suggestions so we can finally merge this:

  1. Move this logic to the last clause of wrap_concatenation. you will need to pass the step and caller as arguments to wrap_concatenation
  2. Only expand and check the arguments if the `__CALLER__` context is match

Reply:
My idea was to do that to be able to check anytime, I changed like you suggested, there is still a problem when you use attributes outside a match context, but I think it's minimum. I think it could be done we remove all the `<>` from Kernel, Macro and Module